crates/bifrost/src/providers/replicated_loglet/rpc_routers.rs

```diff
@@ -55,15 +72,607 @@ impl LogServersRpc {
+            max_inflight_records_in_config,
+            record_permits,
+        }
+    }
+
+    pub async fn append<T: TransportConnect>(
+        &self,
+        loglet: &ReplicatedLoglet<T>,
+        payloads: Arc<[Record]>,
+    ) -> Result<LogletCommit, OperationError> {
+        if payloads.len() > self.max_inflight_records_in_config {
+            let delta = payloads.len() - self.max_inflight_records_in_config;
+            tracing::debug!(
+                "Resizing sequencer in-flight records capacity to allow admission for this batch. \
+                Capacity in configuration is {} and we are adding capacity of {} to it",
+                self.max_inflight_records_in_config,
+                delta
+            );
+            self.record_permits.add_permits(delta);
+        }
+
+        let len = u32::try_from(payloads.len()).expect("batch sizes fit in u32");
+
+        let mut connection = self.connection_manager.get(loglet.networking()).await?;
+
+        let msg = Append {
+            header: CommonRequestHeader {
+                log_id: loglet.log_id(),
+                loglet_id: loglet.params().loglet_id,
+                segment_index: loglet.segment_index(),
+            },
+            payloads: Vec::from_iter(payloads.iter().cloned()),
+        };
+
+        let commit_token = loop {
+            let permits = self
+                .record_permits
+                .clone()
+                .acquire_many_owned(len)
+                .await
+                .unwrap();
+
+            // todo(azmy): avoid copying all records on retry
+            match connection
+                .send(loglet.params().sequencer, permits, msg.clone())
+                .await
+            {
+                Ok(commit_token) => break commit_token,
+                Err(err) => {
+                    match err {
+                        NetworkError::ConnectError(_)
+                        | NetworkError::ConnectionClosed
+                        | NetworkError::Timeout(_) => {
+                            // we retry to re-connect one time
+                            connection = self
+                                .connection_manager
+                                .renew(loglet.networking(), &connection)
+                                .await?;
+                        }
+                        err => return Err(err.into()),
+                    }
+                }
+            }
+        };
+
+        Ok(commit_token)
+    }
+}
+
+/// RemoteSequencerStream represents a single open connection
+/// to remote leader sequencer.
+///
+/// This connection handles all [`Appended`] responses from the remote
+/// sequencer.
+///
+/// If the connection was lost or if any of the commits failed
+/// with a terminal error (like [`Status::Disabled`]) all pending commits
+/// are resolved with an error.
+#[derive(Clone)]
+struct RemoteSequencerConnection {
+    rpc_router: RpcRouter<Append>,
```

Comment:
I don't see a clear reason on why rpc_router should belong here and not in `SequencersRpc`.